pvlib/ivtools/PVsyst_parameter_estimation.py

```diff
@@ -0,0 +1,731 @@
+        of Photovoltaics v5(1), January 2015.
+    [3] A. Mermoud, PV Modules modeling, Presentation at the 2nd PV Performance
+        Modeling Workshop, Santa Clara, CA, May 2013
+    [4] A. Mermoud, T. Lejeuene, Performance Assessment of a Simulation Model
+        for PV modules of any available technology, 25th European Photovoltaic
+        Solar Energy Conference, Valencia, Spain, Sept. 2010
+    [5] C. Hansen, Estimating Parameters for the PVsyst Version 6 Photovoltaic
+        Module Performance Model, Sandia National Laboratories Report
+        SAND2015-8598
+    [6] C. Hansen, Parameter Estimation for Single Diode Models of Photovoltaic
+        Modules, Sandia National Laboratories Report SAND2015-2065
+    [7] C. Hansen, Estimation of Parameters for Single Diode Models using
+        Measured IV Curves, Proc. of the 39th IEEE PVSC, June 2013.
+    """
+    from scipy import optimize
+    import statsmodels.api as sm
+    ee = ivcurves['ee']
+    tc = ivcurves['tc']
+    tck = tc + 273.15
+    isc = ivcurves['isc']
+    voc = ivcurves['voc']
+    imp = ivcurves['imp']
+    vmp = ivcurves['vmp']
+
+    # Cell Thermal Voltage
+    vth = const['k'] / const['q'] * tck
+
+    n = len(ivcurves['voc'])
+
+    # Initial estimate of Rsh used to obtain the diode factor gamma0 and diode
+    # temperature coefficient mugamma. Rsh is estimated using the co-content
+    # integral method.
+
+    pio = np.ones(n)
+    piph = np.ones(n)
+    prsh = np.ones(n)
+    prs = np.ones(n)
+    pn = np.ones(n)
+
+    for j in range(n):
+        current, voltage = rectify_iv_curve(ivcurves['i'][j], ivcurves['v'][j],
+                                            voc[j], isc[j])
+        # initial estimate of Rsh, from integral over voltage regression
+        # [5] Step 3a; [6] Step 3a
+        pio[j], piph[j], prs[j], prsh[j], pn[j] = \
+            estimate_parameters(current, voltage, vth[j] * specs['ns'])
+
+    # Estimate the diode factor gamma from Isc-Voc data. Method incorporates
+    # temperature dependence by means of the equation for Io
+
+    y = np.log(isc - voc / prsh) - 3. * np.log(tck / (const['T0'] + 273.15))
+    x1 = const['q'] / const['k'] * (1. / (const['T0'] + 273.15) - 1. / tck)
+    x2 = voc / (vth * specs['ns'])
+    t0 = np.isnan(y)
+    t1 = np.isnan(x1)
+    t2 = np.isnan(x2)
+    uu = np.logical_or(t0, t1)
+    uu = np.logical_or(uu, t2)
+
+    x = np.vstack((np.ones(len(x1[~uu])), x1[~uu], -x1[~uu] *
+                   (tck[~uu] - (const['T0'] + 273.15)), x2[~uu],
+                   -x2[~uu] * (tck[~uu] - (const['T0'] + 273.15)))).T
+    alpha = np.linalg.lstsq(x, y[~uu])[0]
+
+    gamma_ref = 1. / alpha[3]
+    mugamma = alpha[4] / alpha[3] ** 2
+
+    if np.isnan(gamma_ref) or np.isnan(mugamma) or not np.isreal(gamma_ref) \
+            or not np.isreal(mugamma):
+        badgamma = True
+    else:
+        badgamma = False
+
+    pvsyst = OrderedDict()
+
+    if ~badgamma:
```

Comment:
* This will **always** evaluate to `True`:
```python
>>> ~True
-2

>>> ~False
-1
```
* The Python tilde operator is the bit-complement, so `~0 -> -` and `~1 -> -2`.
* The **NumPy** tilde operator is `not`
* In Python (and NumPy) implicit booleaness is only false for zero:
```python
>>> if ~True: print('yes')
yes

>>> if ~False: print('yes')
yes

>>> if 0: print('yes')

```